debugger/src/tpage.c

```diff
@@ -250,26 +295,38 @@ static void tpage_create_widgets(void)
 	/* target */
 	target_label = gtk_label_new(_("Target:"));
 	target_name = gtk_entry_new ();
+#if GTK_CHECK_VERSION(3, 0, 0)
+	gtk_editable_set_editable(GTK_EDITABLE(target_name), FALSE);
+	target_button_browse = create_stock_button("document-open", _("Browse"));
+#else
 	gtk_entry_set_editable(GTK_ENTRY(target_name), FALSE);
 	target_button_browse = create_stock_button(GTK_STOCK_OPEN, _("Browse"));
+#endif
 	gtk_widget_set_size_request(target_button_browse, BROWSE_BUTTON_WIDTH, 0);
 	g_signal_connect(G_OBJECT(target_button_browse), "clicked", G_CALLBACK (on_target_browse_clicked), NULL);
 
 	/* debugger */
 	debugger_label = gtk_label_new(_("Debugger:")); 
-	debugger_cmb = gtk_combo_box_new_text();
+	debugger_cmb = gtk_combo_box_text_new();
 	modules = debug_get_modules();
 	for (iter = modules; iter; iter = iter->next)
 	{
-		gtk_combo_box_append_text(GTK_COMBO_BOX(debugger_cmb), (gchar*)iter->data);
+		gtk_combo_box_text_append_text(GTK_COMBO_BOX_TEXT(debugger_cmb), (gchar*)iter->data);
 	}
 	g_list_free(modules);
 	gtk_combo_box_set_active(GTK_COMBO_BOX(debugger_cmb), 0);
 
 	/* arguments */
 	args_frame = gtk_frame_new(_("Command Line Arguments"));
+#if GTK_CHECK_VERSION(3, 0, 0)
+	hbox = gtk_scrolled_window_new(
+		gtk_scrollable_get_hadjustment(GTK_SCROLLABLE(args_textview)),
+		gtk_scrollable_get_vadjustment(GTK_SCROLLABLE(args_textview)));
+	gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(hbox), GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC);
```

Comment:
This is weird to be so different between versions, if the view should be in a scrolled window, it probably should be in any case, even if it works OK with GTK2 but shows a GTK3 bug.

Also, `args_textview` seems ti be used before assignment a few lines down.

Reply:
The code can't be the same between GTK+2 and GTK+3 because the GtkTextView (and the GtkTreeView) don't implement the GtkScrollable interface in GTK+2 whereas they do in GTK+3 and shall seemingly be handled with a GtkScrolledWindow to add the scrolling functionality from what I can understand.

However you're absolutely right about the use before assignment, this is fixed.

Reply:
Be it in GTK2 or GTK3, to actually be scrollable you need to put most widgets in a GtkScrolledWindow.  The widgets "supporting scrolling" like GtkTreeView and GtkTextView know how to communicate with the ScrolledWindow.  On GTK3 it's a proper interface; on GTK2 it was signals only.
Basically, what you usually would do is simply create a ScrolledWindow with passing `NULL` as the adjustments, and let GTK do it's thing when you add the child widget.

So, basically you could just use `gtk_scrolled_window_new(NULL, NULL)` followed by `gtk_container_add()` in bot GTK2 and GTK3.

Reply:
Thank you for taking the time to explain this to me. I changed the corresponding code accordingly, as well as some other code which used the same construction for fewer differences between GTK+2 and GTK+3 in debug.c.